yamllint/rules/anchors.py

```diff
@@ -0,0 +1,124 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Use this rule to prevent duplicated anchors and referencing to non-existent anchors.
+
+.. rubric:: Options
+
+* Use ``forbid-unknown-aliases`` to prevent referencing to anchors before assigment or referencing to non-existent anchors.
+* Use ``forbid-duplicated-anchors`` to prevent duplicated anchors.
+
+
+.. rubric:: Examples
+
+#. With ``anchors: {forbid-duplicated-anchors: true}``
+
+   the following code snippet would **PASS**:
+   ::
+
+    clients:
+        jack: &jack_client
+            billing_id: 1234
+        bill: &bill_client
+            billing_id: 5678
+    target_client: *jack_client
+
+   the following code snippet would **FAIL**:
+   ::
+
+    clients:
+        jack: &jack_client
+            billing_id: 1234
+        bill: &jack_client
+            billing_id: 5678
+    target_client: *jack_client
+
+#. With ``anchors: {forbid-unknown-aliases: true}``
+
+   the following code snippet would **PASS**:
+   ::
+
+    address: &address |
+      Williams St. 13
+    target_address: *address
+
+   the following code snippet would **PASS**:
+   ::
+
+    default_address: &address
+      state: South Carolina
+      city:  Barnwell
+    target_address:
+      <<: *address
+      city: Barnaul
+
+   the following code snippet would **FAIL**:
+   ::
+
+    address: &address |
+      Williams St. 13
+    target_address: *wrong_address
+
+
+   the following code snippet would **FAIL**:
+   ::
+
+    default_address: &address
+      state: South Carolina
+      city:  Barnwell
+    target_address:
+      <<: *wrong_address
+      city: Barnaul
+"""
+
+
+from yaml import StreamStartToken, DocumentStartToken, AnchorToken, AliasToken
+
+from yamllint.linter import LintProblem
+
+
+ID = 'anchors'
+TYPE = 'token'
+
+CONF = {'forbid-unknown-aliases': bool,
+        'forbid-duplicated-anchors': bool}
+DEFAULT = {'forbid-unknown-aliases': True,
+           'forbid-duplicated-anchors': False}
+
+
+def check(conf, token, prev, next, nextnext, context):
+    if conf['forbid-unknown-aliases'] or conf['forbid-duplicated-anchors']:
+        # In case of DocumentStartToken `---` is missing
+        if isinstance(token, StreamStartToken):
+            context['anchors'] = []
+        if isinstance(token, DocumentStartToken):
+            context['anchors'] = []
+        elif isinstance(token, AnchorToken):
+            context['anchors'].append(token.value)
+
+    if conf['forbid-unknown-aliases'] and isinstance(token, AliasToken):
+        if token.value not in context['anchors']:
+            yield LintProblem(
+                token.start_mark.line + 1, token.start_mark.column + 1,
+                f'anchor "{token.value}" is used before assignment')
```

Comment:
Should this message also indicate a potentially reference to a non-existent anchor?

Reply:
Yamllint parses files sequencially, so it does not know whether the anchor *is defined later in the file* **or** *is totally missing from the document*.

(Maybe I didn't get your point, in this case please give example of what you mean :))

Reply:
The error message as written right now specifically says something like this:

> anchor "foo-bar" is used before assignment

But, as you said already

>  it does not know whether the anchor is defined later in the file or is totally missing from the document.

That's why I was suggesting that the error message probably could be less specific because it's possible that the anchor hasn't been defined in the entire file. It could be something like:

> anchor "foo-bar" is used before assignment/defined or does not exist

Hope that makes sense.

Reply:
@adrienverge  - Trying to learn the code, as I'm new to this. Would like to contribute if I can.

> Yamllint parses files sequencially, so it does not know whether the anchor is defined later in the file or is totally missing from the document.

Right now problems are reported from each line as tokens are parsed. Is it possible to save them in a global variable instead of reporting right away? Then  if the "end of document token" is reached, report the problems? I didn't see any of the existing rule doing it. Also, can errors from multiple locations be reported? In this example: if an anchor is duplicated 3 times, report the 2nd and 3rd occurrences as problem. Not sure if this is valid:

```python
def check(conf, token, prev, next, nextnext, context):
    # Skipping codes here to keep this snippet short
    yield LintProblem(....)
    yield LintProblem(....)

```

Reply:
> Hope that makes sense.

Yes it does, thanks, I see what you mean now :pray: 
What I like in the sentence `anchor "foo-bar" is used before assignment` is that it's simpler (it counts) and isn't incompatible with "this anchor will actually never be assigned later in the file".
I see this sentence as a good trade-off between simplicity and understandability (compared to `anchor "foo-bar" is used before assignment/defined or does not exist`)


> Right now problems are reported from each line as tokens are parsed. Is it possible to save them in a global variable instead of reporting right away? Then if the "end of document token" is reached, report the problems? I didn't see any of the existing rule doing it.

That looks like a solution, but yamllint is designed to output errors right away, which allows:
\- having them sorted naturally,
\- being performant and avoid consuming too much RAM,
\- not losing past errors if the script crashes at some point.
Changing this would need a major refactor, and I'm not sure this potentially adaptative error message (`... used before assignment` / `... never assigned`) is worth it.

Reply:
I understand what you're saying and it makes sense. I was just asking because I wanted to look at the `forbid-unused-anchors` option. Without checking the entire file, not sure how to verify that an anchor isn't being used. Let me know if you have any suggestions.

Maybe we should discuss it in the feature request issue.

Reply:
See https://github.com/adrienverge/yamllint/issues/395#issuecomment-1423044402